lib/valueflow.cpp

```diff
@@ -618,6 +707,9 @@ static void setTokenValue(Token* tok, const ValueFlow::Value &value, const Setti
 
     // cast..
     if (const Token *castType = getCastTypeStartToken(parent)) {
+        if (astIsPointer(tok) && value.valueType == ValueFlow::Value::INT &&
```

Comment:
I've found that this code fails
```
        check("int foo() {\n"
            "    auto x = getX();\n"
            "    if (x == nullptr)\n"
            "        return 1;\n"
            "    auto y = dynamic_cast<Y*>(x)\n"
            "    if (y == nullptr)\n"
            "        return 2;\n"
            "    return 3;\n"
            "}\n");
        ASSERT_EQUALS("", errout.str());
```

because it incorrectly determines that y must be null. It looks like this code is there to detect and prevent that case (Related to https://trac.cppcheck.net/ticket/9318) however I believe it is failing in this case because the variable in question is defined as auto so astIsPointer returns false (because we don't know if it's a pointer at that stage).

Is it necessary to check if it's a pointer here? I'd have thought we could simply test for dynamic_cast and if we hit it we assume it must be a pointer (or a reference). Perhaps references complicate things?

My solution is
```suggestion
        if (//astIsPointer(tok) && value.valueType == ValueFlow::Value::INT &&
```
but which passes all the unit tests but I'm not sure how comprehensive they are and I could easily be missing something.

Reply:
> which passes all the unit tests but I'm not sure how comprehensive they are and I could easily be missing something.

interesting..

Reply:
#2769 created